Add Mopeka BLE and Mopeka Pro Check BLE Sensor

spbrogan opened the pull request:
# What does this implement/fix? 

Add support for Mopeka Pro Check LP BLE sensor and adds generic component that helps find the BLE MAC address.  
This is a similar pattern to other BLE advertisement based sensors.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <https://github.com/esphome/feature-requests/issues/1445>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1564

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
sensor:
  # Example using 20lb vertical propane tank.
  - platform: mopeka_pro_check
    mac_address: D3:75:F2:DC:16:91
    tank_type: 20LB_V
    temperature:
      name: "Propane test temp"
    level:
      name: "Propane test level"
    distance:
      name: "Propane test distance"
    battery_level:
      name: "Propane test battery level"

  # Custom example - user defined empty / full points
  - platform: mopeka_pro_check
    mac_address: D3:75:F2:DC:16:91
    tank_type: CUSTOM
    custom_distance_full: 40cm
    custom_distance_empty: 10mm
    temperature:
      name: "Propane test temp"
    level:
      name: "Propane test level"
    distance:
      name: "Propane test distance"
    battery_level:
      name: "Propane test battery level"

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

spbrogan committed:
Add support for Mopeka Pro Check LP sensor

spbrogan committed:
Add mopeka_ble component to ease finding a sensor.

Follow pattern used by many EspHome BLE components.
Have a general component to help locate a sensor and have a sensor component for a given sensor type.

spbrogan committed:
Add codeowners to init

spbrogan committed:
Updated actual codeowners file

spbrogan committed:
ci changes

spbrogan committed:
Update test to match syntax change

spbrogan committed:
Update based on coding style

spbrogan committed:
fix syntax

spbrogan committed:
fix clang format issues

spbrogan committed:
rewrite function to match clang format

spbrogan committed:
Update again to single line as formatter is unstable

spbrogan committed:
i really need to get the same clang format working locally

spbrogan commented:
do you want me to resolve the conflict as it is just the test code which should be trivial?  I assume when you are ready to merge you would resolve the conflicts (as more could come up before you choose to merge this).

spbrogan commented:
What needs to happen to get this merged?  Is there a step in the process I am missing?

sdellysse commented:
echoing @spbrogan's comment: I'm building a system with a few ESPHome devices, and having this in-main would be really nice

spbrogan committed:
Update standard US tank min/max to more closely align percentage with app

spbrogan commented:
I just force pushed to the PR.  Only 2 changes.  
1. rebased on latest dev to remove old merge commits
2. updated the defined values for min/max for standard US propane tanks so that the percent calculation more closely aligns with the sensor supplied app.

spbrogan commented:
can this get merged.  I just got notified my documentation PR will be closed if no activity.  Is there something more that is needed?

spbrogan commented:
This has been done since Oct.  Tested, documented, and even requested by other community members.  I can't figure out what needs to happen to get this merged or how to draw the necessary attention to it.  I have noticed plenty of similar new integrations have been merged much faster so again I am looking for some feedback for how this project prioritizes contributions and pull requests.  Should I ask on discord? Do the "maintainers" not have BBQs and thus don't track their propane levels?  :)

jesserockz committed:
Merge branch 'dev' into mopeka_dev

jesserockz commented on esphome/components/mopeka_pro_check/sensor.py:
```diff
@@ -0,0 +1,147 @@
+
+UNIT_MILLIMETER = "mm"
+
+# validators
+_distance_mm = cv.float_with_unit("distance", "(mm|MM)")
+_distance_cm = cv.float_with_unit("distance", "(cm|CM)")
+
+
+def small_distance(value):
+    """small_distance is stored in mm"""
+    try:
+        return _distance_mm(value)
+    except cv.Invalid:
+        pass
+    try:
+        cm = _distance_cm(value)
+        return cm * 10
+    except cv.Invalid:
+        pass
+
+    meters = cv.distance(value)
+    return meters * 100
```
`cv.distance` takes care of all conversions. You just need to multiply the value by 1000 to get mm at the end.

```suggestion
def small_distance(value):
    """small_distance is stored in mm"""
    meters = cv.distance(value)
    return meters * 1000
```

spbrogan replied:
My understanding: cv.distance() only supports meters.  that means a user must enter the values in meters which is awkward.  
https://github.com/esphome/esphome/blob/0ec84be5dacb1a9d0c3cc5c5da5de9a3e9b14df9/esphome/config_validation.py#L848

So the small_distance function allows a user to enter in mm, cm, or meters. 

Am i reading it wrong?

spbrogan replied:
regardless...there is a bug in that line 48 is only doing meters * 100.  So if a user had entered the distance in meters they would not get the right values.  I'll fix that once we have alignment on the above.  Thanks

jesserockz replied:
> Am i reading it wrong?

Yes :smile: 

It means that the base unit and returned value is meters. A user can still put `500mm`, and then the validator will return `0.5` as the actual value (which is why you need the `* 1000`

jesserockz replied:
`float_with_unit` handles prefixes and multiplications on the `m` (for meters)

jesserockz committed:
Update esphome/components/mopeka_pro_check/sensor.py

jesserockz commented on esphome/components/mopeka_pro_check/sensor.py:
```diff
@@ -0,0 +1,132 @@
+TANK_TYPE_CUSTOM = "CUSTOM"
+
+UNIT_MILLIMETER = "mm"
+
```
```suggestion


```

jesserockz committed:
Update esphome/components/mopeka_pro_check/sensor.py

helenclarko commented:
Good to go now then?

jesserockz commented on esphome/components/mopeka_pro_check/sensor.py:
```diff
@@ -0,0 +1,133 @@
+            cv.Required(CONF_MAC_ADDRESS): cv.mac_address,
+            cv.Optional(CONF_CUSTOM_DISTANCE_FULL): small_distance,
+            cv.Optional(CONF_CUSTOM_DISTANCE_EMPTY): small_distance,
+            cv.Optional(CONF_TANK_TYPE, default="20LB_V"): cv.enum(
+                CONF_SUPPORTED_TANKS_MAP, upper=True
+            ),
```
I actually noticed this stand out more when reading the docs PR, should there even be a default here? I think should be required for everyone.
```suggestion
            cv.Required(CONF_TANK_TYPE): cv.enum(
                CONF_SUPPORTED_TANKS_MAP, upper=True
            ),
```

spbrogan committed:
remove default for tank type and make required

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

helenclarko commented:
What does this need now?

jesserockz committed:
Merge branch 'dev' into mopeka_dev

jesserockz committed:
Format with black
